stage1/init/common/mount.go

```diff
@@ -68,131 +69,241 @@ func (m *Mount) Source(podRoot string) string {
 	case "empty":
 		return filepath.Join(common.SharedVolumesPath(podRoot), m.Volume.Name.String())
 	}
-	return "" // We validate in GenerateMounts that it's valid
+	return "" // We validate in RuntimeMounts that it's valid
 }
 
-// GenerateMounts maps MountPoint paths to volumes, returning a list of mounts,
-// each with a parameter indicating if it's an implicit empty volume from a
-// Docker image.
-func GenerateMounts(ra *schema.RuntimeApp, podVolumes []types.Volume, convertedFromDocker bool) ([]Mount, error) {
-	app := ra.App
+// RuntimeMounts represents a mount points generator of the runtime
+// application.
+type RuntimeMounts struct {
+	// Application is the runtime application to generate mount points.
+	Application *schema.RuntimeApp
 
-	var genMnts []Mount
+	// Volumes specifies a list of volumes of the application.
+	Volumes []types.Volume
 
-	vols := make(map[types.ACName]types.Volume)
-	for _, v := range podVolumes {
-		vols[v.Name] = v
+	// DockerImplicit indicates if the resulting mount points should be
+	// marked as implicit empty volumes from a Docker image.
+	DockerImplicit bool
+
+	// processed defines a set of processed mount points, so we don't
+	// have to walk though them multiple times.
+	processed map[string]struct{}
```

Comment:
This seems to indicate that RuntimeMounts is essentially only iterable once.

That makes me a bit uncomfortable - this sort of subtle side-effecting behaviour - I could easily see another developer calling MountsFunc() more than once.

Either RuntimeMounts should error / panic if called twice, or should be non-side-effecting.